src/signature.c

```diff
@@ -157,6 +158,63 @@ GBytes *cms_sign(GBytes *content, const gchar *certfile, const gchar *keyfile, G
 	return res;
 }
 
+gchar* get_pubkey_hash(X509 *cert) {
+		gchar *data = NULL;
+		BIO *bio;
+		unsigned char *ibuf = NULL;
+		unsigned char *buf2;
+		unsigned int len = 0;
+		unsigned int n = 0;
+		unsigned char md[SHA256_DIGEST_LENGTH];
+
+		bio = BIO_new(BIO_s_mem());
+
+		len = i2d_X509_PUBKEY(X509_get_X509_PUBKEY(cert), NULL);
+		buf2 = ibuf = g_malloc(len);
+		if (ibuf) i2d_X509_PUBKEY(X509_get_X509_PUBKEY(cert), &buf2);
+
+		g_assert(buf2 - ibuf == len);
+
+		if (!EVP_Digest(ibuf, len, md, &n, EVP_sha256(), NULL)) {
+			g_warning("Error in EVP_Digest\n");
+			goto out;
+		}
+
+		for (int j = 0; j < (int)n; j++) {
+			BIO_printf(bio, "%02X%c", md[j], (j + 1 == (int)n)
+					? '\0' : ':');
+		}
+
+		BIO_get_mem_data(bio, &data);
+
+		BIO_set_close(bio, BIO_NOCLOSE); /* So BIO_free() leaves BUF_MEM alone */
+		BIO_free(bio);
+out:
+		g_free(ibuf);
+		return data;
+}
+
+gchar** get_pubkey_hashes(STACK_OF(X509) *verified_chain) {
+	GPtrArray *hashes = g_ptr_array_new_full(1, g_free);
```

Comment:
We probably have more than one hash in most cases. 3 certs + 1 NULL pointer seems plausible.

Reply:
Ok, I will init it to 4.